RFC: introduce DebugEventHandlers (fix #650)

sorokin opened the pull request:
OK. Here is my proposed fix for #650.

This problem is solved by introducing an class called
DebugEventHandlers. It holds IDebugEventHandler* and allows removing
when execute_debug_event_handlers is being in progress.

Here is the copy of its description:

DebugEventHandlers is a container of IDebugEventHandler*. It
supports three operations:
1. add -- Adding IDebugEventHandler*
2. remove -- Removing IDebugEventHandler*
3. execute -- Iterating over added handlers and calling them

The reason why a simple standard container is not enough is
that handlers can be added/removed while the list is being
iterated on. Naive implement would cause accessing an invalid
iterator in this case.

In order to support handler removal during iteration this class
keeps an extra bool (inside_execute_) -- whether execution of
handlers is in progress. If it is true then the elements of list
are not erased immediately. Instead they are marked for pending
removal by zeroing their handler pointer.

In order to support handler adding during iteration this class
uses std::list. std::list allows inserting elements without
invalidating iterators to existing elements. Newly added handlers
are inserted into the beginning of the list, so they are not
called on the event that is being executed.

eteran commented:
So I've looked at the issue, which thanks to your very thorough report was very easy to verify 👍.

I think any of your solutions would work, but I believe there is a simpler one...

Make a copy of the container, and it right over that instead in the execute function. (This is Qts approach n their foreach macro). We can avoid the cost of the copy, by using one of Qt's containers since those are copy on write.

sorokin commented:
Thank you. Copying of a container is indeed commonly used practice, and is likely to fix the problem in this case. Also it is much easier to implement, unfortunately it doesn't work in the general case.

Consider we have two handlers A and B and B is located after A in the handler list. Consider that A removed B from the list during the iteration? As the iteration is performed on a copy, B is still in the copy and will be called. It means that a handler is called even after its removal. This is dangerous as often removal is done in destructor and a class can not normally handle events after its removal.

Although this example can look artificial at first, it easily occurs in the programs that use signals widely. The problem with copying is that not only it doesn't support this case, it doesn't allow diagnosing (having an assert) this case occured. The program will just silently fail.

Even if supporting this case is considered unneeded, having an assert on unsupported cases is a good thing. That is why I haven't used copying for fixing this bug.

eteran commented:
That's a reasonable point. I can't look at your PR in-depth just this moment. But I'll be sure to look at it as soon as possible, since this bug should be addressed as quickly as we can.

eteran commented:
Looking at your PR, this is good quality code, and I am convinced it will work. Before I decide to merge, I'd like to explore the different approaches because whenever possible, I usually prefer the simplest solution that gets the job done correctly :-).

I was thinking that if we combine the copy ref counted container approach with a simple check of to see if the element was removed, then we can have the best of both worlds, but have the code be much simpler. The only downside is that it **may** be a little less efficient, but to be honest, I only expect these containers to have an average size of 1-3 elements, so it may be a nearly immeasurable difference. Something like this:

	EVENT_STATUS execute_debug_event_handlers(const std::shared_ptr<IDebugEvent> &e) {

		// if somehow no handler is run, then let's just assume we should stop...
		EVENT_STATUS status = DEBUG_STOP;
	
		// a copy so we can safely mutate as we iterate, QVector<> is ref counted, so this is very cheap on average
		const QVector<IDebugEventHandler*> handlers = g_DebugEventHandlers;

		// loop through all of the handlers, stopping when one thinks that it handled
		// the event
		for(auto it = handlers.begin(); it != handlers.end(); ++it) {
		
			IDebugEventHandler *handler = *it;
		
			// it wasn't removed, so it makes sense to execute it
			if(g_DebugEventHandlers.contains(handler)) {
				status = handler->handle_event(e);
				if(status != DEBUG_NEXT_HANDLER) {
					break;
				}
			}
		}

		// if this assert fails, the bottom handler (which is owned by us) did something terribly wrong :-/
		Q_ASSERT(status != DEBUG_NEXT_HANDLER);
		return status;
	}

In the average case, I would expect this to have basically no cost. additions to the list during execution would obviously simply take effect during the next event.

Thoughts?

eteran commented:
Another option that occurs to me, which is even simpler is to do 3 things:

1. switch to std::list<> so iterators aren't invalidated by insertion
2. state that as a rule, a handler may only remove itself. deleting other handlers is UB.
3. make the execute function look like this:

	EVENT_STATUS execute_debug_event_handlers(const std::shared_ptr<IDebugEvent> &e) {

		// if somehow no handler is run, then let's just assume we should stop...
		EVENT_STATUS status = DEBUG_STOP;

		// loop through all of the handlers, stopping when one thinks that it handled
		// the event
		for(auto it = g_DebugEventHandlers.begin(); it != g_DebugEventHandlers.end(); ) {
	
			// increment before processing, so if it's deleted it's not a problem
			IDebugEventHandler *handler = *it++;
		
			status = handler->handle_event(e);
			if(status != DEBUG_NEXT_HANDLER) {
				break;
			}
		}

		// if this assert fails, the bottom handler (which is owned by us) did something terribly wrong :-/
		Q_ASSERT(status != DEBUG_NEXT_HANDLER);
		return status;
	}

Less robust in that it doesn't support deleting other handlers within a handler, but I'm not sure that is really the most "sane" of circumstances to want to support. I can't really think of a use case which deleting other handlers seems like a good idea ;-)

Anyway, I am curious of your thoughts on this too!

sorokin commented:
In my opinion both of your proposed solution will work. Choosing between these two I like the option 2 (incrementing iterator before deferencing) more, but it is just a personal preference, not the result of rational reasoning.

Speaking of option 1 the only downside I see is its computational complexity which, as you mentioned, doesn't matter if the list is short.

The only thing I would like to add is the debug checks (or asserts) that only the currently executed handler is deleted. These checks can be optional and be enabled only in debug builds. I believe having debug checks like this is important as it makes testing more productive and it simplifies debugging.

I implemented the second approach (with corresponding debug checks). Here you can see the commit: https://github.com/sorokin/edb-debugger/commit/9021493579b9ffa046f7b2f1387bd4d4cbb20d4f, if you like it I can move it into this PR.

eteran commented:
I think the code for option 2 that you've written looks great. My only last question is where is EDB_PRINT_AND_DIE defined? I don't see it anywhere in the github diff but i assume it must be in there somewhere.

10110111 commented:
> My only last question is where is EDB_PRINT_AND_DIE defined?

It's a pre-existing define in `Util.h`.

eteran commented:
@10110111 LOL, somehow I completely forgot that was added!

sorokin commented:
Thank you! EDB_PRINT_AND_DIE was just a random macro I have found when I searched for something to abort the program. 

As the behavior of DebugEventHandlers has changed I would like to rewrite the comment for it. Unfortunately I can not do this right now. I'll update the PR later this evening.

sorokin commented:
I updated the PR. I added a comment for the class describing its behavior. Otherwise the code should be identical to what I linked before (option 2).

As English is not my native language any corrections to the comment and the commit message are welcome.

eteran commented:
merged, looks great. Comments read fine too :-).

I may change the blocks like this:

	auto i = handlers_.begin();
	for (auto end = handlers_.end(); i != end; ++i)
		if (*i == handler)
			break;

into code using `std::find`, but beyond that the code looks very good. Thanks!
